kernel-rs/src/proc.rs

```diff
@@ -223,23 +221,11 @@ impl WaitChannel {
 
     /// Atomically release lock and sleep on waitchannel.
     /// Reacquires lock when awakened.
-    pub unsafe fn sleep<T>(&self, guard: &mut SpinlockGuard<'_, T>) {
-        self.sleep_raw(guard.raw());
-    }
-
-    /// Atomically release lock and sleep on waitchannel.
-    /// Reacquires lock when awakened.
-    pub unsafe fn sleep_sleepable<T>(&self, guard: &mut SleepablelockGuard<'_, T>) {
-        self.sleep_raw(guard.raw());
-    }
-
-    /// Atomically release lock and sleep on waitchannel.
-    /// Reacquires lock when awakened.
-    // TODO(@kimjungwow): lk is not SpinlockGuard yet because
-    // 1. Some static mut variables are still not Spinlock<T> but RawSpinlock
-    // 2. Sleeplock doesn't have Spinlock<T>
-    pub unsafe fn sleep_raw(&self, lk: *const RawSpinlock) {
-        let p: *mut Proc = myproc();
+    /// # Safety
+    ///
+    /// Make sure `lk` is the only lock we currently hold.
```

Comment:
이게 아니면 어떻게 되나요?

Reply:
저도 아직 확실하지는 않지만, 일단은 xv6에서 사용하던 invariant인것 같습니다.
일단, sleep()`내부에서 `sched()을 부르게 되는데, `sched()`함수에 있는 주석에는
```
/// Switch to scheduler.  Must hold only p->lock
/// and have changed proc->state. Saves and restores
/// interrupt_enabled because interrupt_enabled is a property of this
/// kernel thread, not this CPU. It should
/// be proc->interrupt_enabled and proc->noff, but that would
/// break in the few places where a lock is held but
/// there's no process.
```
이라고 나와있어서 이걸 반영한겁니다.

다만, 이건 좀 더 알아보겠습니다!

Reply:
이 조건을 만족하지 않고 `sleep` 을 호출하면 memory safety 에 영향을 주나요? 말씀하신 내용대로면 deadlock 이 문제인데 이것은 safety 와 상관없는 것 같습니다.

Reply:
> 이 조건을 만족하지 않고 sleep 을 호출하면 memory safety 에 영향을 주나요? 말씀하신 내용대로면 deadlock 이 문제인데 이것은 safety 와 상관없는 것 같습니다.

`sched()`와 달리 `scheduler()`에서는 저런 조건이 없는 걸 보면 딱히 그런건 아닌 것 같습니다. 그러면 `unsafe fn sleep()` -> `fn sleep()`으로 바꿀까요? (참고로, `sleep()`을 call하기 전에는 항상 `Cpu->proc`이 non-null이여야 한다는 추가적인 safety 조건이 있기는 한데, 이거는 #258 와 관련해서 나중에 해결해야될 문제 같습니다.)

Reply:
네 말씀하신 대로 `Proc` refactoring 을 전제하면 이 함수는 safe하다고 말할 수 있어 보입니다. `sleep` 은 safe function 으로 바꾸어 주시고, 그 때문에 생기는 unsafe block 에 safety argument 를 추가해 주세요.